### PR TITLE
Remove Lazy GenericRecord implementations

### PR DESCRIPTION
**Pull request by sancar**
The lazy implementation DefaultCompactReader was not
threadsafe. The underlying position is modified at each
read attempt where concurrent reads can not work.

An attempt to solve this problem is here
https://github.com/hazelcast/hazelcast/pull/20040
See commit messages of that to see details.
In short, the code becomes unnecessarily complex.

In this pr, the behaviour is changed so that we
return a completely deserialized GenericRecord
to user via API's like map.get()/EntryProcessor.

The query path still uses the lazy implementation. We do
not deserialize whole data still.

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java [hunk 1]
```diff
@@ -775,8 +915,8 @@ public void writeJsonFormattedField(StringBuilder stringBuilder, AbstractGeneric
         };
         ALL[FieldKind.PORTABLE.getId()] = new FieldKindBasedOperations() {
             @Override
-            public Object readObject(GenericRecord genericRecord, String fieldName) {
-                return ((InternalGenericRecord) genericRecord).getObject(fieldName);
+            public Object readAsLeafObjectOnQuery(InternalGenericRecord genericRecord, String fieldName) {
+                return genericRecord.getObject(fieldName);
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java:919]: So, we are trying to read an object out of InternalGenericRecord, right? That feels a bit weird, why do we need it?

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/DeserializedGenericRecord.java [hunk 2]
```diff
@@ -553,4 +554,210 @@ protected Object getClassIdentifier() {
...
+        if (array == null || array.length <= index) {
+            return null;
+        }
+        return array[index];
+    }
+
+    @Nullable
+    @Override
+    public Character getCharFromArray(@Nonnull String fieldName, int index) {
+        char[] array = getArrayOfChar(fieldName);
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/DeserializedGenericRecord.java:592]: we could throw UnsupportedOperationException here

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java (hunk 1 above)
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java:919]: This is used from Query when the user's query ends with an object. 
`SELECT * FROM map WHERE x.y.z.employee > ? `

In this case, we have to construct `Employee` class to be able to compare on the server. This method is called in this case.

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java [hunk 3]
```diff
@@ -41,7 +41,12 @@
     static {
         ALL[FieldKind.BOOLEAN.getId()] = new FieldKindBasedOperations() {
             @Override
-            public Boolean readObject(GenericRecord genericRecord, String fieldName) {
+            public Object readAsLeafObjectOnQuery(InternalGenericRecord genericRecord, String fieldName) {
+                return genericRecord.getBoolean(fieldName);
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java:45]: I don't have strong feeling here. Making it separate felt much more obvious. I will change it.

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/InternalGenericRecord.java [hunk 4]
```diff
@@ -31,13 +31,34 @@
  * Additionally to GenericRecord, this one has more methods to be used in Query.
  *
  * @see GenericRecordQueryReader
+ * InternalGenericRecord should be constructed in a deserialized manner to gain maximum performance on Query
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/InternalGenericRecord.java:34]: I tried to mention the fact that those implementations does not deserialize the objects completely. I can reword this part with, 
```
InternalGenericRecord implementations should not deserialize the passed content completely in their constructor when they are created to be used in query.
```
Please suggest more if that is not clear as well.

**Comment by sancar** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/InternalGenericRecord.java [hunk 5]
```diff
@@ -31,13 +31,35 @@
...
  * read*FromArray methods will return `null`
  * 1. if the array is null or empty
  * 2. there is no data at given index. In other words, the index is bigger than the length of the array
  */
 public interface InternalGenericRecord extends GenericRecord {
 
+    /**
+     * @param fieldName the name of the field
+     * @return the value of the field
+     * @throws HazelcastSerializationException if the field name does not exist in the class definition or
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/InternalGenericRecord.java:48]:
```suggestion
     * @throws HazelcastSerializationException if the field name does not exist in the class definition/schema or
```

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/InternalGenericRecord.java [hunk 6]
```diff
@@ -31,13 +31,35 @@
...
+     *                                         the type of the field does not match the one in the class definition/schema.
+     */
+    @Nullable
+    InternalGenericRecord getInternalGenericRecord(@Nonnull String fieldName);
+
+    /**
+     * @param fieldName the name of the field
+     * @return the value of the field
+     * @throws HazelcastSerializationException if the field name does not exist in the class definition or
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/InternalGenericRecord.java:57]:
```suggestion
     * @throws HazelcastSerializationException if the field name does not exist in the class definition/schema or
```

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializer.java [hunk 7]
```diff
@@ -260,14 +232,28 @@ private CompactSerializableRegistration getOrCreateRegistration(String typeName)
...
+    private GenericRecord readGenericRecord(BufferObjectDataInput input, Schema schema, boolean schemaIncludedInBinary) {
+        DefaultCompactReader reader = new DefaultCompactReader(this, input, schema, null, schemaIncludedInBinary);
+        Collection<FieldDescriptor> fields = schema.getFields();
+        DeserializedSchemaBoundGenericRecordBuilder builder = new DeserializedSchemaBoundGenericRecordBuilder(schema);
+        for (FieldDescriptor fieldDescriptor : fields) {
+            String fieldName = fieldDescriptor.getFieldName();
+            FieldKind fieldKind = fieldDescriptor.getKind();
+            builder.write(fieldName, reader.readAny(fieldName), fieldKind);
+        }
+        return builder.build();
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializer.java:244]: There seem to be some unnecessary checks in these two calls. Maybe we can put a TODO comment here to improve things in the future

File: hazelcast/src/main/java/com/hazelcast/nio/serialization/AbstractGenericRecord.java [hunk 8]
```diff
@@ -28,7 +29,7 @@
 /**
  * Implementation of GenericRecord interface to give common equals and hashcode implementation
  */
-public abstract class AbstractGenericRecord implements GenericRecord {
+public abstract class AbstractGenericRecord implements InternalGenericRecord {
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/nio/serialization/AbstractGenericRecord.java:32]: I see that this class is extended in lots of places including the ones we give to the user. Is that necessary? If so, there are some unnecessary `implements InternalGenericRecord` statements on some of the classes that extend this(or one of its childs)

File: hazelcast/src/main/java/com/hazelcast/query/impl/getters/CompactGetter.java [hunk 9]
```diff
@@ -33,7 +33,7 @@
     @Override
     Object getValue(Object target, String fieldPath) throws Exception {
         InternalGenericRecord record;
-        if (target instanceof CompactInternalGenericRecord) {
+        if (target instanceof CompactGenericRecord) {
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/query/impl/getters/CompactGetter.java:36]: I see that we will use this getter if `targetObject instanceof CompactGenericRecord` (from `Extractors#L148`). So, how come this if will return false we hit the else branch?

File: hazelcast/src/main/java/com/hazelcast/nio/serialization/AbstractGenericRecord.java (hunk 8 above)
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/nio/serialization/AbstractGenericRecord.java:32]: When in memory format is Object , the stored object is DeserializedGenericRecord. 
And the query needs to work on it. `GenericRecordQueryReader` excepts only `InternalGenericRecord`. 
Therefore all implementations needs to be `InternalGenericRecord`

**Review by mdumandag — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/InternalGenericRecord.java [hunk 10]
```diff
@@ -31,13 +31,35 @@
  * Additionally to GenericRecord, this one has more methods to be used in Query.
  *
  * @see GenericRecordQueryReader
+ * InternalGenericRecord implementations should not deserialize the passed content completely in their constructor when
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/InternalGenericRecord.java:34]: Updated at https://github.com/hazelcast/hazelcast/pull/20043/commits/4ea64ed4c51cf4bd5544c9da7bd9b4a8d3b890e6

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/DeserializedGenericRecord.java [hunk 11]
```diff
@@ -557,4 +558,206 @@ protected Object getClassIdentifier() {
...
+    @Nullable
+    @Override
+    public Character getCharFromArray(@Nonnull String fieldName, int index) {
+        throw new UnsupportedOperationException("Compact format does not support reading a char field");
+    }
+
+    @Nullable
+    @Override
+    public Double getFloat64FromArray(@Nonnull String fieldName, int index) {
+        double[] array = getArrayOfFloat64(fieldName);
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/DeserializedGenericRecord.java:605]: This is the `DeserializedGenericRecord`. All the fields are already deserialized at the construction. 
And this line does not make a copy.

**Review by ihsandemir — APPROVED** [review]

**Pull request merged**